src/session/sessions.ts

```diff
@@ -0,0 +1,284 @@
+    if (value) {
+      Cookies.set(this.refreshTokenName, value, {
+        /* Expire time is set to 6 days */
+        expires: 6,
+      });
+    } else {
+      Cookies.remove(this.refreshTokenName);
+    }
+  }
+
+  //> Methods
+  /**
+   * Get a valid session token. If there is not the session will be refreshed.
+   *
+   * @returns {Promise<string | undefined>} The session token if set
+   */
+  async upToDateToken(): Promise<string | undefined> {
+    let token = super.token;
+
+    /* Refresh token if there is none */
+    if (!token) {
+      await this.refresh();
+
+      token = super.token;
+    }
+
+    return token;
+  }
+
+  /**
+   * Send a query to the endpoint.
+   *
+   * @param {string} token A authentication token
+   * @param {DocumentNode} data A DocumentNode with a query
+   * @param {object} variables A object with variables
+   */
+  async send(token: string, data: DocumentNode, variables?: object) {
+    let headers = {
+      authorization: token,
+    };
+
+    return this.ep.sendQuery(data, variables, headers);
+  }
+
+  /**
+   * Begin session.
+   *
+   * @param {string} user A User defined by username and password
+   * @returns {Promise<any>} A UserData object
+   */
+  async begin(user?: User): Promise<any> {
+    let response;
+
+    if (!user && this.refreshToken) {
+      /* Refresh token and retrieve a new refreshToken if necessary */
+      this.refresh();
+    } else {
+      if (!user) {
+        /* Authenticate anonymous user */
+        response = await this.tasks.auth.anon();
+      } else {
+        /* Authenticate real user */
+        response = await this.tasks.auth.nonanon(user);
+      }
+
+      /* Set tokens */
+      this.token = response.data?.auth.token;
+      this.refreshToken = response.data?.auth.refreshToken;
+
+      return response.data?.auth.user;
+    }
+
+    /* Get user data */
+    response = await this.tasks.user.whoami();
+
+    return response.data?.whoami;
+  }
+
+  /**
+   * Refreshes a session based on its history.
+   *
+   * @description When there is no token the refresh task is called. When there
+   *              is no token and refresh token session begin as equivalent to
+   *              an anonymous login is called.
+   */
+  async refresh() {
+    if (!this.token) {
+      if (this.refreshToken) {
+        let response = await this.tasks.auth.refresh();
+
+        this.token = response.data?.refresh.token;
+        this.refreshToken = response.data?.refresh.refreshToken;
+      } else {
+        /* No token and refresh token present, start anonymous login */
+        await this.begin();
+      }
+    }
+  }
+
+  /**
+   * Ends a session.
+   *
+   * @description The token and refresh token are revoked and deleted
+   */
+  async end() {
+    /* Revoke token if it is set */
+    if (this.refreshToken !== "") {
+      let response = await this.tasks.auth.revoke();
+
+      //#DEBUG TSID1
+      console.log("TID-1(REVOKE)", response.data?.revoke.revoked);
```

Comment:
Should console.logs be active on potential release code?
I'd suggest to implement a norm on console.logs, so they can be activated and deactivated by replacement very fast.
I suggest using `//~console.log();` - so on project cloning you can replace `//~` with ` `  and on deployment or master merge you can replace `console.log` with `//~console.log` to keep the integrity.

Reply:
The norm already exists //#TSID if I remember correctly.